fuzzers/libafl-fuzz/src/executor.rs

```diff
@@ -0,0 +1,185 @@
+            opt.executable = full_bin_path;
+            bin_path = &opt.executable;
+        } else {
+            return Err(Error::illegal_argument(format!(
+                "Program '{}' not found or not executable",
+                opt.executable.display()
+            )));
+        }
+    } else {
+        bin_path = &opt.executable;
+        #[cfg(target_os = "linux")]
+        {
+            if opt.nyx_mode {
+                if !bin_path.is_symlink() && bin_path.is_dir() {
+                    let config_file = bin_path.join("config.ron");
+                    if !config_file.is_symlink() && config_file.is_file() {
+                        return Ok(());
+                    }
+                }
+                return Err(Error::illegal_argument(
+                    format!(
+                        "Directory '{}' not found, or is a symlink or is not a nyx share directory",
+                        bin_path.display()
+                    )
+                    .as_str(),
+                ));
+            }
+        }
+    }
+    metadata = bin_path.metadata()?;
+    let is_reg = !bin_path.is_symlink() && !bin_path.is_dir();
+    let bin_size = metadata.st_size();
+    let is_executable = metadata.permissions().mode() & 0o111 != 0;
+    if !is_reg || !is_executable || bin_size < 4 {
+        return Err(Error::illegal_argument(format!(
+            "Program '{}' not found or not executable",
+            bin_path.display()
+        )));
+    }
+    if opt.skip_bin_check
+        || opt.use_wine
+        || opt.unicorn_mode
+        || (opt.qemu_mode && opt.qemu_custom_bin)
+        || (opt.forkserver_cs && opt.cs_custom_bin)
+        || opt.non_instrumented_mode
+    {
+        return Ok(());
+    }
+
+    let file = File::open(bin_path)?;
+    let mmap = unsafe { MmapOptions::new().map(&file)? };
+
+    // check if it's a shell script
+    if mmap[0..1] == [0x43, 0x41] {
+        // TODO: finish error message
+        return Err(Error::illegal_argument(
+            "Oops, the target binary looks like a shell script.",
+        ));
+    }
+    #[cfg(target_os = "macos")]
+    {
+        // check if the binary is an ELF file
+        if mmap[0..4] != [0x7f, 0x45, 0x4c, 0x46] {
+            return Err(Error::illegal_argument(format!(
+                "Program '{}' is not an ELF binary",
+                bin_path.display()
+            )));
+        }
+
+        #[cfg(not(target_arch = "arm"))]
+        {
+            if (mmap[0] != 0xCF || mmap[1] != 0xFA || mmap[2] != 0xED)
+                && (mmap[0] != 0xCA || mmap[1] != 0xFE || mmap[2] != 0xBA)
+            {
+                return Err(Error::illegal_argument(format!(
+                    "Program '{}' is not a 64-bit or universal Mach-O binary",
+                    bin_path.display()
+                )));
+            }
+        }
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    let check_instrumentation = !opt.qemu_mode
```

Comment:
frida and unicorn mode are also supported on macos/freebsd/whatever fwiw

Reply:
I've simplified the check